felix/bpf/mock/multiversion/upgrade.go

```diff
@@ -0,0 +1,184 @@
+// Copyright (c) 2022 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mock
+
+import (
+	"fmt"
+
+	"github.com/projectcalico/calico/felix/bpf"
+	"github.com/projectcalico/calico/felix/bpf/cachingmap"
+	v2 "github.com/projectcalico/calico/felix/bpf/mock/multiversion/v2"
+	v3 "github.com/projectcalico/calico/felix/bpf/mock/multiversion/v3"
+	v4 "github.com/projectcalico/calico/felix/bpf/mock/multiversion/v4"
+	v5 "github.com/projectcalico/calico/felix/bpf/mock/multiversion/v5"
+)
+
+func MapV2(mc *bpf.MapContext) bpf.Map {
+	b := mc.NewPinnedMap(v2.MockMapParams)
+	b.(*bpf.PinnedMap).UpgradeFn = Upgrade
+	return b
+}
+
+func MapV3(mc *bpf.MapContext) bpf.Map {
+	b := mc.NewPinnedMap(v3.MockMapParams)
+	b.(*bpf.PinnedMap).UpgradeFn = Upgrade
+	return b
+}
+
+func MapV4(mc *bpf.MapContext) bpf.Map {
+	b := mc.NewPinnedMap(v4.MockMapParams)
+	b.(*bpf.PinnedMap).UpgradeFn = Upgrade
+	return b
+}
+
+func MapV5(mc *bpf.MapContext) bpf.Map {
+	b := mc.NewPinnedMap(v5.MockMapParams)
+	b.(*bpf.PinnedMap).UpgradeFn = Upgrade
+	return b
+}
+
+func NewKeyV2(k uint32) v2.Key {
+	return v2.NewKey(k)
+}
+
+func NewKeyV3(k uint32) v3.Key {
+	return v3.NewKey(k)
+}
+
+func NewKeyV4(k uint32) v4.Key {
+	return v4.NewKey(k)
+}
+
+func NewKeyV5(k uint32) v5.Key {
+	return v5.NewKey(k)
+}
+
+func NewValueV2(k uint32) v2.Value {
+	return v2.NewValue(k)
+}
+
+func NewValueV3(k uint32) v3.Value {
+	return v3.NewValue(k)
+}
+
+func NewValueV4(k uint32) v4.Value {
+	return v4.NewValue(k)
+}
+func NewValueV5(k uint32) v5.Value {
+	return v5.NewValue(k)
+}
```

Comment:
I think you can just import the vN packages in the test and place this in the test and use vX.NewY() :shrug: